### PR TITLE
[Core][ParallelUtilities] adding GlobalLock

### PR DESCRIPTION
**Pull request by philbucher**
**Description**
This is a draft PR for the `GlobalLock` which we can use to replace `omp critical`

more on this in the upcoming parallelization discussion @KratosMultiphysics/technical-committee

### CODE REVIEW

File: kratos/utilities/parallel_utilities.cpp [hunk 1]
```diff
@@ -24,8 +25,12 @@
 #include "includes/lock_object.h"
 
 
-namespace Kratos
-{
+namespace Kratos {
+
+namespace {
+    std::mutex GlobalLockInitializerMutex;
```
Comment (philbucher) [inline kratos/utilities/parallel_utilities.cpp]: this is the important part
If this works (which I think it does as it is only declared and used once here)
then we can use it for initializing the `GlobalLock` ...

File: kratos/utilities/parallel_utilities.cpp [hunk 2]
```diff
@@ -106,22 +111,35 @@ int ParallelUtilities::InitializeNumberOfThreads()
...
 }
 
+LockObject& ParallelUtilities::GetGlobalLock()
+{
+    if (!mspGlobalLock) {
+        const std::lock_guard<std::mutex> scope_lock(GlobalLockInitializerMutex);
+        if (!mspGlobalLock) {
+            static LockObject global_lock;
+            mspGlobalLock = &global_lock;
+        }
```
Comment (philbucher) [inline kratos/utilities/parallel_utilities.cpp]: ... here

**Comment by philbucher** [comment]
I stumbled across [call_once](https://www.cplusplus.com/reference/mutex/call_once/) which seems to be the ideal solution for this

**Comment by philbucher** [comment]
from the docs:
```Calls fn passing args as arguments, unless another thread has already executed (or is currently executing) a call to call_once with the same flag.```

**Review by pooyan-dadvand — APPROVED** [review]
@KratosMultiphysics/technical-committee approves it. Thanks!

**Pull request merged**